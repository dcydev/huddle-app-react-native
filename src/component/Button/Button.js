import * as React from 'react';
import { ActivityIndicator, TouchableOpacity, View } from 'react-native';

import Text from '/component/Text';

import useTimeout from '/hook/useTimeout';

import {
  mainStyles,
  stylesByType,
  stylesBySize,
  disabledButton,
} from './Button.styles';

import globalStyles from '/styles';

/**
 * @typedef {('primary'|'secondary'|'ghost'|'danger')} ButtonTypes
 * @typedef {('small'|'medium'|'large')} ButtonSize
 *
 * @param {Object} props
 * @param {ButtonSize} [props.size]
 * @param {ButtonTypes} [props.type]
 * @param {boolean} [props.disabled]
 * @param {function} [props.onPress]
 * @param {string} props.text
 * @param {string} [props.style]
 * @param {string} [props.textStyle]
 */
const Button = ({
  size = 'medium',
  type = 'primary',
  onPress,
  text,
  disabled = false,
  style,
  textStyle,
  loading,
  testID,
}) => {
  let Wrapper = TouchableOpacity;

  const touchable = !(loading || disabled);

  const handlePress = useTimeout(onPress);

  if (!touchable) {
    Wrapper = View;
  }

  const loadingColour = type === 'ghost'
    ? globalStyles.palette.deepBlue
    : globalStyles.palette.white;

  return (
    <Wrapper
      onPress={(touchable) ? handlePress : undefined}
      testID={testID}

      style={[
        mainStyles.button,
        stylesBySize[size].button,
        stylesByType[type].button,
        disabled && disabledButton.button,
        style,
      ]}
    >
      {!loading && (
        <Text
          style={[
            stylesByType[type].text,
            stylesBySize[size].text,
            disabled && disabledButton.text,
            textStyle,
          ]}
          numberOfLines={1}
          weight={type === 'ghost' ? 'regular' : 'medium'}
          lineHeight={24}
        >
          {text}
        </Text>
      )}
      {!!loading && (
        <ActivityIndicator
          color={loadingColour}
          testID="loading"
        />
      )}
    </Wrapper>
  );
};

export default Button;
