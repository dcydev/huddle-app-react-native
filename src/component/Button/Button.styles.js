import { StyleSheet } from 'react-native';
import globalStyles from '/styles';

export const mainStyles = StyleSheet.create({
  button: {
    borderRadius: 8,
    flex: 0,
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
  },
});

export const stylesByType = {
  default: StyleSheet.create({}),
  primary: StyleSheet.create({
    button: {
      backgroundColor: globalStyles.palette.deepBlue,
    },
    text: {
      color: globalStyles.palette.white,
    },
  }),
  secondary: StyleSheet.create({
    button: {
      backgroundColor: globalStyles.palette.grey01,
    },
    text: {
      color: globalStyles.palette.white,
    },
  }),
  ghost: StyleSheet.create({
    button: {
      backgroundColor: globalStyles.palette.white,
      borderColor: globalStyles.palette.grey02,
      borderWidth: 1,
    },
    text: {
      color: globalStyles.palette.deepBlue,
    },
  }),
  danger: StyleSheet.create({
    button: {
      backgroundColor: globalStyles.palette.orange,
    },
    text: {
      color: globalStyles.palette.white,
    },
  }),
};

export const stylesBySize = {
  small: StyleSheet.create({
    button: {
      height: 32,
      paddingHorizontal: 12,
      borderRadius: 4,
    },
    text: {
      fontSize: 12,
    },
  }),
  medium: StyleSheet.create({
    button: {
      height: 44,
      paddingHorizontal: 24,
    },
    text: {
      fontSize: 14,
    },
  }),
  large: StyleSheet.create({
    button: {
      height: 52,
      paddingHorizontal: 24,
      alignSelf: 'stretch',
    },
    text: {
      fontSize: 16,
    },
  }),
};

export const disabledButton = StyleSheet.create({
  button: {
    backgroundColor: globalStyles.palette.grey04,
  },
  text: {
    color: globalStyles.palette.black,
    opacity: 0.4,
  },
});
