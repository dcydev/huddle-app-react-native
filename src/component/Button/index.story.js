import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { storiesOf } from '@storybook/react-native';
import { StyleSheet, View } from 'react-native';

import globalStyles from '/styles';
import Button from './index';

const containerStyle = StyleSheet.create({
  container: {
    padding: globalStyles.padding.sm,
  },
});

const Container = ({ children }) => <View style={containerStyle.container}>{children}</View>;

storiesOf('Buttons', module).add('primary button', () => (
  <Container>
    <Button text="Primary button" />
  </Container>
));
