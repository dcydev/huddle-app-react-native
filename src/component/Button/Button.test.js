import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react-native';
import Button from './Button';

describe('Button', () => {
  const buttonText = 'Press Me';

  let component;
  let props;
  let mockOnPress;

  beforeEach(() => {
    mockOnPress = jest.fn();

    component = () => {
      const utils = render(
        <Button
          type="primary"
          text={buttonText}
          onPress={mockOnPress}
          {...props}
        />,
      );

      return {
        ...utils,
      };
    };
  });

  afterEach(() => {
    jest.resetModules();
    cleanup();
  });

  describe('when using default button', () => {
    it('renders correctly', () => {
      const { baseElement } = component();
      expect(baseElement).toMatchSnapshot();
    });

    it('shows correct text', () => {
      const { getByText } = component();
      expect(getByText(buttonText)).toBeTruthy();
    });

    it('doesn\'t show ActivityIndicator component', () => {
      const { queryByTestId } = component();
      expect(queryByTestId('loading')).toBeNull();
    });

    describe('when pressed', () => {
      beforeEach(() => {
        const { queryByText } = component();
        fireEvent.press(queryByText(buttonText));
      });

      it('calls onPress prop', () => {
        expect(mockOnPress).toBeCalledTimes(1);
      });
    });
  });

  describe('when using disabled button', () => {
    beforeAll(() => {
      props = {
        disabled: true,
      };
    });

    it('renders correctly', () => {
      const { baseElement } = component();
      expect(baseElement).toMatchSnapshot();
    });

    it('doesn\'t show ActivityIndicator component', () => {
      const { queryByTestId } = component();
      expect(queryByTestId('loading')).toBeNull();
    });

    it('shows correct text', () => {
      const { getByText } = component();
      expect(getByText(buttonText)).toBeTruthy();
    });

    describe('when pressed', () => {
      beforeEach(() => {
        const { queryByText } = component();
        fireEvent.press(queryByText(buttonText));
      });

      it('doesn\'t call onPress prop', () => {
        expect(mockOnPress).not.toBeCalled();
      });
    });
  });

  describe('when using loading prop', () => {
    beforeAll(() => {
      props = {
        loading: true,
      };
    });

    it('renders correctly', () => {
      const { baseElement } = component();

      expect(baseElement).toMatchSnapshot();
    });

    it('shows ActivityIndicator component', () => {
      const { getByTestId } = component();
      expect(getByTestId('loading')).toBeTruthy();
    });

    it('doesn\'t show text', () => {
      const { queryByText } = component();
      expect(queryByText(buttonText)).toBeNull();
    });

    describe('when pressed', () => {
      beforeEach(() => {
        const { queryByTestId } = component();
        fireEvent.press(queryByTestId('loading'));
      });

      it('doesn\'t call onPress prop', () => {
        expect(mockOnPress).not.toBeCalled();
      });
    });
  });
});
