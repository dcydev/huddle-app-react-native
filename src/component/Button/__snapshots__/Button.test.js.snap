// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button when using default button renders correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 8,
            "flex": 0,
            "flexDirection": "row",
            "justifyContent": "center",
          },
          Object {
            "height": 44,
            "paddingHorizontal": 24,
          },
          Object {
            "backgroundColor": "#203054",
          },
          false,
          undefined,
        ]
      }
    >
      <Text
        allowFontScaling={false}
        numberOfLines={1}
        style={
          Array [
            undefined,
            Object {
              "color": "#203054",
              "fontFamily": "Lato",
              "fontSize": 14,
              "fontWeight": "500",
              "lineHeight": 24,
            },
            false,
            Array [
              Object {
                "color": "#fff",
              },
              Object {
                "fontSize": 14,
              },
              false,
              undefined,
            ],
          ]
        }
      >
        Press Me
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`Button when using disabled button renders correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 8,
            "flex": 0,
            "flexDirection": "row",
            "justifyContent": "center",
          },
          Object {
            "height": 44,
            "paddingHorizontal": 24,
          },
          Object {
            "backgroundColor": "#203054",
          },
          Object {
            "backgroundColor": "#f6f6f6",
          },
          undefined,
        ]
      }
    >
      <Text
        allowFontScaling={false}
        numberOfLines={1}
        style={
          Array [
            undefined,
            Object {
              "color": "#203054",
              "fontFamily": "Lato",
              "fontSize": 14,
              "fontWeight": "500",
              "lineHeight": 24,
            },
            false,
            Array [
              Object {
                "color": "#fff",
              },
              Object {
                "fontSize": 14,
              },
              Object {
                "color": "#000",
                "opacity": 0.4,
              },
              undefined,
            ],
          ]
        }
      >
        Press Me
      </Text>
    </View>
  </View>
</View>
`;

exports[`Button when using loading prop renders correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 8,
            "flex": 0,
            "flexDirection": "row",
            "justifyContent": "center",
          },
          Object {
            "height": 44,
            "paddingHorizontal": 24,
          },
          Object {
            "backgroundColor": "#203054",
          },
          false,
          undefined,
        ]
      }
    >
      <ActivityIndicator
        animating={true}
        color="#fff"
        hidesWhenStopped={true}
        size="small"
        testID="loading"
      />
    </View>
  </View>
</View>
`;
